src/client/dfuse/ops/read.c

```diff
@@ -12,159 +12,160 @@
-	      struct fuse_file_info *fi)
+dfuse_cb_read(fuse_req_t req, fuse_ino_t ino, size_t len, off_t position, struct fuse_file_info *fi)
 {
 	struct dfuse_obj_hdl		*oh = (struct dfuse_obj_hdl *)fi->fh;
 	struct dfuse_projection_info	*fs_handle = fuse_req_userdata(req);
 	const struct fuse_ctx		*fc = fuse_req_ctx(req);
-	struct fuse_bufvec		fb = {};
 	void				*buff;
 	int				rc;
-	size_t				buff_len = len;
-	bool				skip_read = false;
+	size_t				buff_len  = len;
+	bool				mock_read = false;
 	bool				readahead = false;
-	bool				async = false;
 	struct dfuse_event		*ev = NULL;
 
+	if (ino != oh->doh_ie->ie_stat.st_ino) {
+		DFUSE_TRA_ERROR(oh, "Inconsistent inode number %lu %lu", ino,
+				oh->doh_ie->ie_stat.st_ino);
+		D_GOTO(err, rc = EIO);
+	}
+
 	D_ALLOC_PTR(ev);
 	if (ev == NULL)
 		D_GOTO(err, rc = ENOMEM);
 
 	DFUSE_TRA_UP(ev, oh, "event");
 
-	DFUSE_TRA_INFO(ev, "%#zx-%#zx requested pid=%d",
-		       position, position + len - 1, fc->pid);
+	DFUSE_TRA_DEBUG(ev, "%#zx-%#zx requested pid=%d", position, position + len - 1, fc->pid);
 
-	if (oh->doh_ie->ie_truncated &&
-	    position + len < oh->doh_ie->ie_stat.st_size &&
-		((oh->doh_ie->ie_start_off == 0 &&
-			oh->doh_ie->ie_end_off == 0) ||
-			position >= oh->doh_ie->ie_end_off ||
-			position + len <= oh->doh_ie->ie_start_off)) {
+	if (oh->doh_ie->ie_truncated && position + len < oh->doh_ie->ie_stat.st_size &&
+	    ((oh->doh_ie->ie_start_off == 0 && oh->doh_ie->ie_end_off == 0) ||
+	     position >= oh->doh_ie->ie_end_off || position + len <= oh->doh_ie->ie_start_off)) {
 		off_t pos_ra = position + len + READAHEAD_SIZE;
 
 		DFUSE_TRA_DEBUG(oh, "Returning zeros");
-		skip_read = true;
+		mock_read = true;
 
 		if (pos_ra <= oh->doh_ie->ie_stat.st_size &&
-		    ((oh->doh_ie->ie_start_off == 0 &&
-				oh->doh_ie->ie_end_off == 0) ||
-				(position >= oh->doh_ie->ie_end_off ||
-					pos_ra <= oh->doh_ie->ie_start_off))) {
+		    ((oh->doh_ie->ie_start_off == 0 && oh->doh_ie->ie_end_off == 0) ||
+		     (position >= oh->doh_ie->ie_end_off || pos_ra <= oh->doh_ie->ie_start_off))) {
 			readahead = true;
 		}
-	} else if (oh->doh_caching &&
-		len < (1024 * 1024) &&
-		oh->doh_ie->ie_stat.st_size > (1024 * 1024)) {
+	} else if (oh->doh_caching && len < (1024 * 1024) &&
+		   oh->doh_ie->ie_stat.st_size > (1024 * 1024)) {
 		/* Only do readahead if the requested size is less than 1Mb and
 		 * the file size is > 1Mb
 		 */
 
 		readahead = true;
 	}
 
-	if (readahead) {
-		if (!fs_handle->dpi_info->di_wb_cache)
-			buff_len += READAHEAD_SIZE;
-	} else {
-		if (!skip_read) {
-			rc = daos_event_init(&ev->de_ev,
-					     fs_handle->dpi_eq, NULL);
-			if (rc != -DER_SUCCESS)
-				D_GOTO(err, rc = daos_der2errno(rc));
-
-			ev->de_req = req;
-			ev->de_complete_cb = dfuse_cb_read_complete;
-			async = true;
-		}
-	}
+	if (fs_handle->dpi_info->di_wb_cache)
+		readahead = false;
+
+	if (readahead)
+		buff_len += READAHEAD_SIZE;
```

Comment:
Would we want to align the readahead size in some way?

Reply:
This isn't a change in behaviour, but possibly yes.  Getting the readahead triggers right is probably an important thing to do and more work on this area would help but it's outside the scope of this PR.

Reply:
Makes sense, and I agree.